### ISSUE TITLE
Scout scanning and live reconstruction

### ISSUE THREAD

**Issue by eibhinwhite** [PlusToolkit/PlusLib#665]
Hi,
I'm currently attempting to 3 dimensionally reconstruct ultrasound images using the plus remote in slicer. I have the ultrasound live image and optical tracker loaded into slicer and they are working fine. However, when I click on scout scan it shows a red x beside it and does not allow me to open the ROI eye on live reconstruction. I have code within my loaded ultrasound images for live reconstruction. I'm wondering what my problem may be ? Thank you

**Comment by eibhinwhite** [comment]
<PlusConfiguration version="2.1">

  <DataCollection StartupDelaySec="1.0" >
    <DeviceSet
      Name="PlusServer: Ultrasonix US Eibhin"
      Description="Broadcasting ultrasound images acquired from the Ultrasonix system through OpenIGTLink. Probe is spatially calibrated for any imaging depth. If PlusServer does not run on the Ultrasonix PC then update the IP attribute in the Device element with the Ultrasonix PC's IP address. Ascension3DG sensors should be plugged in to the Ascension3DG DriveBay mounted on Ultrasonix US in the following order from to leftmost slot (Transducer 1) to the right: 1 Probe, 2 Reference, 3 Stylus." />
	
   <Device
      Id="VideoDevice"
      Type="SonixVideo" 
      AcquisitionRate="30" 
      LocalTimeOffsetSec="-0.2976"
      IP="192.168.0.4"
      AutoClipEnabled="TRUE"
      ImageToTransducerTransformName="ImageToReference" >
      <DataSources>
        <DataSource Type="Video" Id="Video" PortName="B" PortUsImageOrientation="UN"  />
      </DataSources>
      <OutputChannels>
        <OutputChannel Id="VideoStream" VideoDataSourceId="Video"/>
      </OutputChannels>
    </Device>
	
    <Device 
      Id="TrackedVideoDevice" 
      Type="VirtualMixer" >
      <InputChannels>
        <InputChannel Id="VideoStream" />
      </InputChannels>
      <OutputChannels>
        <OutputChannel Id="TrackedVideoStream"/>
      </OutputChannels>
    </Device>
   
   <Device
      Id="CaptureDevice"
      Type="VirtualCapture"
      BaseFilename="Recording.igs.mhd"
      EnableCapturingOnStart="FALSE"
      RequestedFrameRate="15" >
      <InputChannels>
        <InputChannel Id="TrackedVideoStream" />
      </InputChannels>
    </Device>
	
    <Device
      Id="VolumeReconstructorDevice"
      Type="VirtualVolumeReconstructor"
      OutputVolDeviceName="Volume_Reference"
      EnableReconstruction="FALSE" >
      <InputChannels>
        <InputChannel Id="TrackedVideoStream" />
      </InputChannels>
      <VolumeReconstruction
        ImageCoordinateFrame="Image" ReferenceCoordinateFrame="Reference"
        CompoundingMode="MEAN" Interpolation="LINEAR"
        Optimization="FULL" NumberOfThreads="2"
        ClipRectangleOrigin="0 0" ClipRectangleSize="820 616" PixelRejectionThreshold="1"
        OutputSpacing="0.5 0.5 0.5"
        FillHoles="ON" >
        <HoleFilling>
          <HoleFillingElement Type="GAUSSIAN" Size="5" Stdev="0.6667" MinimumKnownVoxelsRatio="0.50001" />
          <HoleFillingElement Type="STICK" StickLengthLimit="9" NumberOfSticksToUse="1" />
        </HoleFilling>
      </VolumeReconstruction>
    </Device>
  </DataCollection>

  <CoordinateDefinitions>
    
    <Transform From="Transducer" To="Probe"
      Matrix="
        0.0018    0.9477   -0.0175   14.8449
       -1.0000    0.0016   -0.0052   15.0061
       -0.0052    0.0166    0.9998    0.1638
             0         0         0    1.0000" />
  </CoordinateDefinitions>
  
   <PlusOpenIGTLinkServer 
    MaxNumberOfIgtlMessagesToSend="1" 
    MaxTimeSpentWithProcessingMs="50" 
    ListeningPort="18946" 
    SendValidTransformsOnly="true" 
    OutputChannelId="VideoStream" > 
    <DefaultClientInfo> 
      <MessageTypes> 
        <Message Type="IMAGE" />
		<Message Type="TRANSFORM" />
      </MessageTypes>
	  
      <TransformNames> 

        <Transform Name="ProbeToReference" />

        <Transform Name="StylusToReference" />

        <Transform Name="ReferenceToTracker" />

      </TransformNames>
      <ImageNames>
        <Image Name="Transducer" EmbeddedTransformToFrame="Transducer" />
      </ImageNames>
    </DefaultClientInfo>
  </PlusOpenIGTLinkServer>
  
</PlusConfiguration>

**Comment by eibhinwhite** [comment]
![image](https://user-images.githubusercontent.com/57176240/74944118-6bd99e80-53ed-11ea-8ca1-15b9b818054f.png)
![image](https://user-images.githubusercontent.com/57176240/74944157-7bf17e00-53ed-11ea-8772-e134170e57aa.png)
![image](https://user-images.githubusercontent.com/57176240/74944236-a2afb480-53ed-11ea-9ca8-ab7bd0218a20.png)
![image](https://user-images.githubusercontent.com/57176240/74944261-ad6a4980-53ed-11ea-99d3-170ecc179da9.png)

**Comment by eibhinwhite** [comment]
This is the code I am currently using for Ultrasound images and live reconstruction. I am launching the optical tracker through another plus server seperately.

**Comment by Sunderlandkyl** [comment]
I don't see an optical tracker in your config file?
If you hover over the red x, there should be a tooltip with the error.
Could you also upload your Plus log file?

**Comment by eibhinwhite** [comment]
Hi, yes theres no optical tracking in this as I am launching the optical
tracking through another plus server simultaneously, would this perhaps be
the problem? I can use the optical tracking within slicer but perhaps
that's why the reconstructions not working ? Also, when I hover over the
red X of the scout scan nothing appears, but if I click live reconstruction
after the scout scanning and hover over that red x, it says 'no ROI set
yet. start with scout scan', even though it won't allow me to open the ROI
eye.

**Comment by Sunderlandkyl** [comment]
You need tracking information in the same PlusServer to complete the volume reconstruction.

I'll take a look at why the tooltip doesn't appear.

**Comment by eibhinwhite** [comment]
Ok thank you this is really helpful! I'm wondering is there any possibility if I send in my code for optical tracking, would you have any idea of how to add it to my code above to allow it all to go through one server? Sorry I'm just very new to coding and finding this a bit difficult! Thanks again for your help.
![image](https://user-images.githubusercontent.com/57176240/74950864-ab0ced00-53f7-11ea-8bf7-d8b22a063886.png)
![image](https://user-images.githubusercontent.com/57176240/74951185-ce379c80-53f7-11ea-86ac-5dd9c77a48f4.png)
![image](https://user-images.githubusercontent.com/57176240/74951580-e3143000-53f7-11ea-9c10-bab3641e5cc7.png)

**Comment by Sunderlandkyl** [comment]
If you add all of your devices to the same config file and feed the ultrasound and optical tracking data into the same virtual mixer device, it should work.
I would look at this config file as a template: https://github.com/PlusToolkit/PlusLibData/blob/master/ConfigFiles/PlusDeviceSet_Server_Ultrasonix_C5-2_Ascension3DG_calibrated.xml

**Comment by eibhinwhite** [comment]
Ok thank you. So I've tried this kind of template and it doesn't seem to be working. Sorry for so many questions just would really like to get this up and running. Would you mind having a look?
![image](https://user-images.githubusercontent.com/57176240/74954343-dfce7380-53fa-11ea-9c8a-73c2d5217468.png)
![image](https://user-images.githubusercontent.com/57176240/74954383-ee1c8f80-53fa-11ea-899d-a22cbaf4157c.png)
![image](https://user-images.githubusercontent.com/57176240/74954410-fbd21500-53fa-11ea-9514-8bb48369ff08.png)
![image](https://user-images.githubusercontent.com/57176240/74954471-13110280-53fb-11ea-94b0-514b0f89e4eb.png)
![image](https://user-images.githubusercontent.com/57176240/74954507-2328e200-53fb-11ea-8b01-a476035441d1.png)

**Comment by eibhinwhite** [comment]
![image](https://user-images.githubusercontent.com/57176240/74971127-7b6ddd00-5417-11ea-836b-29d2f02803be.png)
![image](https://user-images.githubusercontent.com/57176240/74971163-8a548f80-5417-11ea-9018-2935f2a1065e.png)
![image](https://user-images.githubusercontent.com/57176240/74971218-9fc9b980-5417-11ea-9c62-bd72b6b9abe9.png)
![image](https://user-images.githubusercontent.com/57176240/74971251-afe19900-5417-11ea-8abb-8f56dd1cbf91.png)
![image](https://user-images.githubusercontent.com/57176240/74971287-bff97880-5417-11ea-802a-afb35f4d4cea.png)
So I've cleaned it up a bit, however, when I launch it through the Plus win32 Ultrasonix 5.7 Plus Server, it says 'unknown device type:Mmf video'. Is there a way around this?

**Comment by Sunderlandkyl** [comment]
You could try the PlusApp-2.9.0.X-Win32 package. Alternatively, you could swap the Sonix video device for an [OpenIGTLinkVideo](http://perk-software.cs.queensu.ca/plus/doc/nightly/user/DeviceOpenIGTLinkVideo.html) device, and use it to connect to a separate Ultrasonix 5.7 Plus Server running only the SonixVideo.

**Comment by eibhinwhite** [comment]
So would I include this openIGTLinkVideo code with the code I have above,
replacing the sonix video, and then launch the sonix video on its own
through the Ultrasonix 5.7 server? Thanks for your help.

**Comment by eibhinwhite** [comment]
![image](https://user-images.githubusercontent.com/57176240/75023533-dd6a2900-548f-11ea-830d-825cee8c650d.png)
So I've tried launching my original one through the Plus 2.9.x win 32 Server and this shows.

**Comment by eibhinwhite** [comment]
![image](https://user-images.githubusercontent.com/57176240/75024882-1dcaa680-5492-11ea-9b01-6f3e94dabc06.png)
So I've tried running the code using the Plus 2.9.X win32 and this is whats showing.

**Comment by eibhinwhite** [comment]
Hi,
Just following from my previous issue, I've made some adjustments. So this is what I have done exactly.
First, I put together a config file including the optical tracking, volume reconstruction and also an IGTLink with the same port as the Sonix video (which is run seperately). I then launched this file through the Plus Server (2.8 version) win32 on my laptop.


[PlusServer_TRY.TXT](https://github.com/PlusToolkit/PlusLib/files/4239714/PlusServer_TRY.TXT)




 I then downloaded the Plus Server (version 2.8) Ultrasonix 5.7 win 32 to my Ultrasonix machine and launched a config file including just the Sonix video device. 
[PlusServer_SONIXONLY.TXT](https://github.com/PlusToolkit/PlusLib/files/4239715/PlusServer_SONIXONLY.TXT)

Basically, I've set the IGTLink Server port to match the sonix video listening port and set the IGTLink Server Address to that of the Ultrasonix machine's IP address.
 I am getting an error in the error log which I think is based off the transform names. I'm wondering would you mind taking a look at my config files and see if you think theres some error perhaps in my code? Or even just your thoughts on my method, whether you think this is the right way to go about it. Thank you for your help so far!

**Comment by eibhinwhite** [comment]
Hi,
Just following on from my previous issue, I've made some adjustments. So this is what I have done exactly.
First, I put together a config file including the optical tracking, volume reconstruction and also an IGTLink with the same port as the Sonix video (which is run seperately). I then launched this file through the Plus Server (2.8 version) win32 on my laptop.
[PlusServer_TRY.TXT](https://github.com/PlusToolkit/PlusLib/files/4239723/PlusServer_TRY.TXT)

I then downloaded the Plus Server (version 2.8) Ultrasonix 5.7 win 32 to my Ultrasonix machine and launched a config file including just the Sonix video device.
[PlusServer_SONIXONLY.TXT](https://github.com/PlusToolkit/PlusLib/files/4239725/PlusServer_SONIXONLY.TXT)

Basically, I've set the IGTLink Server port to match the sonix video listening port and set the IGTLink Server Address to that of the Ultrasonix machine's IP address.
When launching both at the same time they both have successfu connections, however I am getting an error in the error log and it's not really working.
This is the error log for the Ultrasonix machine launching the Sonix video:  
[022120_160955_PlusLog.txt](https://github.com/PlusToolkit/PlusLib/files/4239720/022120_160955_PlusLog.txt)
And this is the error log for the config file launched on my laptop:
[022120_153240_PlusLog.txt](https://github.com/PlusToolkit/PlusLib/files/4239722/022120_153240_PlusLog.txt)

I'm wondering would you mind taking a look at my config files and see if you think theres some error perhaps in my code (maybe the transforms)? Or even just your thoughts on my method, whether you think this is the right way to go about it. Thank you for your help so far!

**Comment by lassoan** [comment]
We are moving questions and discussions from the issue tracker to https://github.com/PlusToolkit/PlusLib/discussions. If you have remaining questions related to this topic, create a new discussion there. Thank you!
